proto/cosmos/staking/v1beta1/genesis.proto

```diff
@@ -38,6 +39,93 @@ message GenesisState {
+
+  // epoch number at the time of genesis
+  int64 epoch_number = 10;
+
+  // Current logic flow summary:
+  // Current queues: "unbonding validator", "unbonding delegator", “redelegation"
+  // Current Queue management
+  // EndBlocker -> BlockValidatorUpdates -> DequeueAllMatureUBDQueue
+  // EndBlocker -> BlockValidatorUpdates -> DequeueAllMatureRedelegationQueue
+  // EndBlocker -> BlockValidatorUpdates -> UnbondAllMatureValidators 
+  // EndBlocker -> BlockValidatorUpdates -> CompleteUnbonding -> RemoveUnbondingDelegation
+  // EndBlocker -> BlockValidatorUpdates -> CompleteRedelegation -> RemoveRedelegation
+  // BeginBlocker -> HandleValidatorSignature -> Jail
+  // BeginBlocker -> HandleValidatorSignature -> Slash
+
+  // Target logic flow summary:
+  // Epoch -> BlockValidatorUpdates -> DequeueAllMatureUBDQueue
+  // Epoch -> BlockValidatorUpdates -> DequeueAllMatureRedelegationQueue
+  // Epoch -> BlockValidatorUpdates -> UnbondAllMatureValidators 
+  // Epoch -> BlockValidatorUpdates -> CompleteUnbonding -> RemoveUnbondingDelegation
+  // Epoch -> BlockValidatorUpdates -> CompleteRedelegation -> RemoveRedelegation
+  // BeginBlocker -> HandleValidatorSignature -> Jail
+  // BeginBlocker -> HandleValidatorSignature -> Slash
+
+  // Current Msgs management
+  // MsgUnjail -> instant validator update on condition met
+  // MsgCreateValidator -> instant validator creation
+  // MsgEditValidator -> instant validator update
+  // MsgBeginRedelegate -> instant validator update and queue redelegation
+  // MsgUndelegate -> instant pool balance update and queue undelegate
+  // MsgDelegate -> instant delegation update
+  // MsgWithdrawValidatorCommission -> instant
+  // MsgWithdrawDelegatorReward -> instant
+
+  // Target Msgs management
+  // MsgUnjail -> queue validator update on condition met (BufferedMsgUnjailQueue)
+  // MsgCreateValidator -> queue validator creation on condition met (BufferedMsgCreateValidatorQueue)
+  // MsgEditValidator -> queue validator edit on condition met (BufferedMsgEditValidatorQueue)
+  // MsgDelegate -> queue delegate (BufferedMsgDelegateQueue)
+  // MsgBeginRedelegate -> queue redelegation (BufferedMsgRedelegationQueue) => move tokens between validators on epoch => After 3 weeks time, it should automatically remove redelegation entry for completion even though it's nott the end of epoch
+  // MsgUndelegate -> queue undelegation (BufferedMsgUndelegateQueue) => move tokens to NotBondedPool and start unbonding period on epoch => After 3 weeks time, it should automatically unbond even though it’s not the end of epoch
+  // MsgWithdrawValidatorCommission -> instant
+  // MsgWithdrawDelegatorReward -> instant
+
+  // The flow for an unbonding process would be:
+
+  // 1. Submit MsgUnbond which adds it to DelegationChangeQueue
+  // 2. Wait for end of Epoch
+  // 3. Execute "BeginUnbonding", this adds it to UnbondingQueue
+  // 4. Wait till end of Unbonding Period (3 weeks)
```

Comment:
Separately, we should also probably change the unbonding period to be (3 weeks - epoch time) once we do the delayed epoching

Reply:
It's already implemented like that. Unbonding just start after epoching message execution.